Added Router

sxv2721 opened the pull request:
Added react router dom, and update my components to respond to the current URL. On click the description and images expand, and the url is changed to reflect the current selected launch. Also updated boolean naming conventions to be more descriptive.

fixes #8

sxv2721 committed:
Added Router

Added react router dom, and update my components to respond to the current URL. On click the description and images expand, and the url is changed to reflect the current selected launch. Also updated boolean naming conventions to be more descriptive.

bursteways commented on src/components/DateInput/index.js:
```diff
@@ -6,43 +6,43 @@ export class DateInput extends React.Component {
     }
 
-    startChange = (e) => {
+    startDateChange = (e) => {
```
A preferred convention for these handlers is to set a prefix. You will see some developers use `on`, others will user `handle`. Both are acceptable. So for these functions, consider changing them to something like:

```
onChangeDate = (e, key) => this.setState({ [key]: e.target.value });
```

then in the implementation, you can use the same function, but pass it a key:
```
onChange={(e) => this.onChangeDate(e, 'startDate')}
```

bursteways commented on src/components/DateInput/index.js:
```diff
@@ -6,43 +6,43 @@ export class DateInput extends React.Component {
     }
     componentDidMount(){
-        this.props.getLaunches(this.state.start, this.state.end);
+        this.props.getLaunches(this.state.startDate, this.state.endDate);
```
A common convention is to destructure when you have multiple keys off of an object. So you might consider:

```
const { startDate, endDate } = this.state;
this.props.getLaunches(startDate, endDate);
```

That way you're not repeating `this.state`.

bursteways commented on src/components/DateInput/index.js:
```diff
@@ -6,43 +6,43 @@ export class DateInput extends React.Component {
-                    max={this.state.end}
+                    value={this.state.startDate}
+                    onChange={this.startDateChange}
+                    max={this.state.endDate}
```
I'm curious about `max`. Do you need to specify this?

sxv2721 replied:
min and max limit the input options to only valid date ranges, so the drop down menu won't list invalid dates.

https://www.w3schools.com/tags/att_input_max.asp

bursteways commented on src/components/DateInput/index.js:
```diff
@@ -6,43 +6,43 @@ export class DateInput extends React.Component {
+                    onChange={this.endDateChange}
+                    min={this.state.startDate}
                 />
                 <button onClick={(e) => {
```
In order to clean this up, I would recommend abstracting this out into a class level function.

```
// In class
this.onSubmit = (e) => {
  const { startDate, endDate } = this.state
  e.preventDefault();
  this.props.getLaunches(startDate, endDate);
};
// In render
<button
  onClick={this.onSubmit}
  className="dateSubmit"
>
Submit
</button>
```

bursteways replied:
I see an opportunity to reuse `this.props.getLaunches(startDate, endDate);` Maybe you can make a function for this too.

bursteways commented on src/components/Launch/components/LaunchImage/index.js:
```diff
@@ -4,36 +4,46 @@ import "./styles.scss";
-        this.state = {//need to add derivedstatefromprops(nextProps, prevState)
-            showImages: this.props.imgBool || false
+        this.state = {
+            isShowingImages: this.props.isShowingImages || false
```
Good change, prefixing with `is` 👍

bursteways replied:
This should be `props.isShowingImages`.

bursteways commented on src/components/Launch/components/LaunchImage/index.js:
```diff
@@ -4,36 +4,46 @@ import "./styles.scss";
                     alt="mission patch"
                     className="launchImg"></img>
-                {this.props.links.flickr_images.length > 0 &&
+                {(this.state.isShowingImages && this.props.isCurrentRoute===false) &&
```
When you have a really long confusing statement like this, it's a good idea to pull this out into a constant at the "render" level.

```
const { isCurrentRoute } = this.props;
const { isShowingImages } = this.state;
const shouldShowButton = isShowingImages && isCurrentRoute === false;
```

Consider destructuring the rest of `this.props` and `this.state` below.

bursteways commented on src/components/Launch/components/LaunchImage/index.js:
```diff
@@ -4,36 +4,46 @@ import "./styles.scss";
                         return <img src={img}
                             alt={"flickr " + index}
                             className="flickrImg"
                             key={"flickr " + index}></img>;
```
React prefers self-closing tags when there are no children. So for this image, you can use `<img ... />`.

bursteways commented on src/components/Launch/index.js:
```diff
@@ -29,38 +29,53 @@ export class Launch extends React.Component {
                 <button className="favoriteButton"
                     onClick={(e) => {//dispatch favorites actions here.
-                        if (!this.props.heart) {
+                        if (!this.props.isFavorited) {
```
This `onClick` should be pulled out into a class method.

It's _always_ a best practice to _not_ inline functions within a render. The reason is because React "re-renders" very often and that means that each time the component re-renders you are creating a new function every time.

If you pull this out into a class method, React will "cache" the instance of this method and it will save you on some function creation/rendering.

bursteways commented on src/components/Launch/index.js:
```diff
@@ -29,38 +29,53 @@ export class Launch extends React.Component {
+}
+
+/*
+<button className="descButton" onClick={() => {
```
Do you need these comments?

bursteways commented on src/components/Launches/index.js:
```diff
@@ -1,20 +1,26 @@
-                        return (<Launch
-                            key={"Launch" + index}
+                        return (
+                        <NavLink to={"/"+launch.mission_name} className="navLink" activeClassName="activeLink" key={"Launch" + index}>
```
Consider using `string-interpolation` for `to`. Google it 😉 it's part of ES6.

bursteways commented on src/components/Launches/index.js:
```diff
@@ -1,20 +1,26 @@
                             data={launch}
-                            heart={this.props.favorites.includes(launch)}
-                        />)
+                            isFavorited={this.props.favorites.includes(launch)}
```
Don't forget to destructure just outside the render.

bursteways commented on src/components/Launch/index.js:
```diff
@@ -29,38 +29,53 @@ export class Launch extends React.Component {
         return (
             <section className="launch" >
-                <LaunchImage links={data.links} imgBool={this.state.imgBool} />
+                <Route path={"/" + data.mission_name}
```
String interpolation!

I need to take a closer look at the router - this looks a bit different than I remember. We can look at this together in person.

sxv2721 committed:
Merge branch 'master' into refactor-8

sxv2721 committed:
created routed missions

When you click on a launch card it now routes to that launch full screen and shows images. Removed LaunchImage. Updated tests.

sxv2721 committed:
Merge branch 'refactor-8' of https://github.com/sxv2721/SpaceX-Launches-App into refactor-8
